### PR TITLE
[MM-29053] prevent using format for logs

### PR DESCRIPTION
**Pull request by attiss**
#### Summary

This PR prevents using the `--format` flag and the `MMCTL_FORMAT=json` setting for the `mmctl logs` subcommand.

#### Ticket Link

Fixes https://github.com/mattermost/mattermost-server/issues/15693

### CODE REVIEW

**Comment by attiss** [comment]
Tried to add a nice unit test for this into `logs_test.go`, but unfortunately as `format` is a global flag, this wasn't enough.

```go
	s.Run("Error when using format flag", func() {
		cmd := &cobra.Command{}
		cmd.Flags().String("format", "json", "")

		data, err := testLogsCmdF(s.client, cmd, []string{})

		s.Require().Error(err)
		s.Require().Len(data, 0)
	})
```

(Even if I added the check to `logsCmdF` function.)

I'd appreciate some help on this.

**Comment by attiss** [comment]
```
$ ./mmctl logs --help
Display logs in a human-readable format. The "--format" flag cannot be used here.
```

Manual tests:
```
$ ./mmctl logs
2020-10-01T19:59:50.1685932+02:00 info app/server.go:216 Server is initializing...
2020-10-01T19:59:50.1687715+02:00 info app/web_hub.go:83 number_of_hubs=4 Starting websocket hubs
2020-10-01T19:59:50.1694715+02:00 warn app/server.go:225 Sentry reporting is enabled, but SENTRY_DSN is not set. Disabling reporting.
....
```

```
$ ./mmctl logs --format json
Error: format cannot be used
...
```

```
$ MMCTL_FORMAT=json ./mmctl logs
Error: format cannot be used
...
```

**Comment by attiss** [comment]
Trying to add a unit test for checking whether the `--format` flag was set, but I didn't find a way for that yet.

A unit test like this is not sufficient:
```go
	s.Run("Error when using format flag", func() {
		cmd := &cobra.Command{}
		cmd.Flags().String("format", "json", "")

		data, err := testLogsCmdF(s.client, cmd, []string{})

		s.Require().Error(err)
		s.Require().Len(data, 0)
	})
```

Will continue to dig this up, any help is appreciated!

File: commands/logs.go [hunk 1]
```diff
@@ -11,13 +11,16 @@ import (
...
+	"github.com/spf13/viper"
 
 	"github.com/mattermost/mmctl/client"
+	"github.com/mattermost/mmctl/printer"
 )
 
 var LogsCmd = &cobra.Command{
 	Use:   "logs",
 	Short: "Display logs in a human-readable format",
+	Long:  "Display logs in a human-readable format. The \"--format\" flag cannot be used here.",
```
Comment (mgdelacroix) [inline commands/logs.go:23]: Suggesting a small change for clarity:

```suggestion
	Long:  "Display logs in a human-readable format. As the logs format depends on the server, the \"--format\" flag cannot be used with this command.",
```

File: commands/logs.go [hunk 2]
```diff
@@ -28,6 +31,10 @@ func init() {
 }
 
 func logsCmdF(c client.Client, cmd *cobra.Command, args []string) error {
+	if cmd.Flags().Changed("format") || viper.GetString("format") == printer.FormatJSON {
+		return errors.New("format cannot be used")
```
Comment (mgdelacroix) [inline commands/logs.go:35]: Suggesting a change on the text for clarity

```suggestion
		return errors.New("the \"--format\" flag cannot be used with this command")
```

**Review by mgdelacroix — CHANGES_REQUESTED** [review]
Looking really good @attiss, thanks for the contribution!! :rocket: 

Added a couple of change requests

**Comment by mgdelacroix** [comment]
> Trying to add a unit test for checking whether the `--format` flag was set, but I didn't find a way for that yet.

You can set the flag as changed with:

```go
cmd := &cobra.Command{}
cmd.Flags().String("format", "json", "")
cmd.Flags().Lookup("format").Changed = true
```

**Comment by attiss** [comment]
Thanks for the review @mgdelacroix!

Fixed the messages. I decided to separate the flag-set check and the environment variable check in order to provide more meaningful error responses. (If you prefer the previous implementation, just say and I'm happy to change it back.)

Thanks for the idea on the unit test, I'd never found it! 😄 

---

```
$ ./mmctl logs --help
Display logs in a human-readable format. As the logs format depends on the server, the "--format" flag cannot be used with this command.
```

Manual tests:
```
$ ./mmctl logs
2020-10-01T22:50:01.3744278+02:00 info expirynotify/worker.go:85 worker=ExpiryNotify job_id=n3jrppqqsbdu8m151e394fc9ph Worker: Job is complete
2020-10-01T23:00:01.4789832+02:00 info expirynotify/worker.go:85 worker=ExpiryNotify job_id=h3t6wzkc5ifnjb3dguky34ffyy Worker: Job is complete
2020-10-01T23:00:01.8869336+02:00 info product_notices/worker.go:85 worker=ProductNotices job_id=j5xp63co7ff6fpw8rtfdwnd1rc Worker: Job is complete
....
```

```
$ ./mmctl logs --format json
Error: the "--format" flag cannot be used with this command
...
```

```
$ MMCTL_FORMAT=json ./mmctl logs
Error: json formatting cannot be applied on the output of this command
...
```

File: commands/logs.go (hunk 1 above)
Comment (attiss) [reply commands/logs.go]: fixed

File: commands/logs.go (hunk 2 above)
Comment (attiss) [reply commands/logs.go]: fixed

**Review by ethervoid — APPROVED** [review]
Thank you very much for your contribution! This PR looks good for me :)

File: commands/logs.go [hunk 3]
```diff
@@ -28,6 +31,12 @@ func init() {
 }
 
 func logsCmdF(c client.Client, cmd *cobra.Command, args []string) error {
+	if cmd.Flags().Changed("format") {
+		return errors.New("the \"--format\" flag cannot be used with this command")
+	} else if viper.GetString("format") == printer.FormatJSON {
+		return errors.New("json formatting cannot be applied on the output of this command")
```
Comment (mgdelacroix) [inline commands/logs.go:37]: I'm happy making this distinction, good thinking! My only concern is that the error message doesn't explicitly point to the environment variable as it is now, and as sometimes those can be somehow obscure, I'd only suggest to change this to state that the value comes from the environment variable:

```suggestion
		return errors.New("json formatting cannot be applied on this command. Please check the value of \"MMCTL_FORMAT\"")
```

**Review by mgdelacroix — CHANGES_REQUESTED** [review]

File: commands/logs.go (hunk 3 above)
Comment (attiss) [reply commands/logs.go]: Fair point, fixed!

```
$ MMCTL_FORMAT=json ./mmctl logs                                                              
Error: json formatting cannot be applied on this command. Please check the value of "MMCTL_FORMAT"
...
```

**Review by mgdelacroix — APPROVED** [review]
LGTM!! Thanks a lot for applying the changes @attiss!! :confetti_ball:

**Comment by hanzei** [comment]
@attiss Heads up that the test `TestMmctlUnitSuite/TestLogsCmd/Error_when_setting_json_format_with_environment_variable` did fail on CI. Could you please take a look?

**Comment by attiss** [comment]
whoops, my bad, forgot to update the expected error message in the unit test :)

fixed, thanks @hanzei

File: mattermost-server [hunk 4]
```diff
@@ -0,0 +1 @@
+/home/attiss/git/github.com/mattermost/mattermost-server
```
Comment (hanzei) [inline mattermost-server:1]: Did you intend to add this file?

Comment (attiss) [reply mattermost-server]: whoops again, just added that accidentally, thanks for noticing! fixed

**Review by lindy65 — APPROVED** [review]
Thanks @attiss @mgdelacroix - Looks good to me 👍

**Pull request merged**